web/app.js

```diff
@@ -1542,6 +1542,33 @@ const PDFViewerApplication = {
     };
     internalEvents.set("updatefromsandbox", updateFromSandbox);
 
+    const visitedPages = new Set();
+    const pageOpen = async ({ pageNumber }) => {
+      let actions = null;
+      if (!visitedPages.has(pageNumber)) {
+        // Get actions only the first time the page is visited
```

Comment:
The comment is misleading here, since you don't actually prevent multiple fetches of the same data until it's resolved at least *once*.
Furthermore, there's also generally speaking a risk that this data resolves *after* the page has been "closed" , meaning that the event-ordering is not guaranteed. Is that not an issue?

Also, you shouldn't generally need to re-fetch the `PDFPageProxy` instance since it will usually be available already here.